File: graph.rb
require 'benchmark'
require 'async'

MEGABYTE = 2**20
FILENAME = './soc-LiveJournal1.txt'.freeze
FILE_SIZE = File.size(FILENAME).to_f / MEGABYTE
CHUNK_SIZE = (FILE_SIZE / 16).floor * MEGABYTE

def profile_memory
  memory_usage_before = `ps -o rss= -p #{Process.pid}`.to_i
  yield
  memory_usage_after = `ps -o rss= -p #{Process.pid}`.to_i

  used_memory = ((memory_usage_after - memory_usage_before) / 1024.0).round(2)
  puts "Memory usage: #{used_memory} MB"
end

def profile_time
  time_elapsed = Benchmark.realtime do
    yield
  end

  puts "Time: #{time_elapsed.round(2)} seconds"
end

def profile_gc
  GC.start
  before = GC.stat(:total_freed_objects)
  yield
  GC.start
  after = GC.stat(:total_freed_objects)

  puts "Objects Freed: #{after - before}"
end

def profile
  profile_memory do 
    profile_time do 
      profile_gc do
        yield
      end
    end 
  end 
end

profile do
  ractors = []
  File.open(FILENAME).each(nil, CHUNK_SIZE) do |chunk|
    require 'pry'; binding.pry
    ractors << Ractor.new(chunk) do |new_chunk|
      graph = {}
      new_chunk.split("\n").each do |line|
        next if line[0] == '#'

        from_node, to_node = line.split
        graph[from_node.to_i] = (graph[from_node.to_i] || []).append(to_node.to_i)
      end
      graph
    end
  end
  result = ractors.map(&:take)
  puts result[0]
end

File: graph.go
package main

import (
    "bufio"
    "fmt"
    "log"
    "os"
    "strings"
    "strconv"
    "time"
)

func main () {
  start := time.Now()
  file, err := os.Open("soc-LiveJournal1.txt")
  if err != nil {
    log.Fatal(err);
    return
  }
  defer file.Close()

  scanner := bufio.NewScanner(file)
  graph := make(map[int64][]int64)
  for scanner.Scan() {
    line := scanner.Text()
    if string(line[0]) == "#" {
      continue
    }
    values := strings.Split(line, "\t")
    fromNode, _ := strconv.ParseInt(values[0], 10, 64)
    toNode, _ := strconv.ParseInt(values[1], 10, 64)
    _, valExists := graph[fromNode]
    if valExists {
      graph[fromNode] = append(graph[fromNode], toNode)
    } else {
      graph[fromNode] = []int64{}
    }
  }
  fmt.Println(graph[0])
  elapsed := time.Since(start)
  fmt.Println("Elapsed time: ", elapsed)
}
